src/components/organisms/LocalReplace.tsx

```diff
@@ -0,0 +1,438 @@
+  onReplaceToggle,
+  onReplaceQueryChange,
+  onReplacementFinished,
+  onFocusSearchInput,
+  onReplaceClose,
+  onUpdateSearchOptions,
+}: LocalReplaceProps) => {
+  const replaceTextAreaRef = useRef<HTMLTextAreaElement>(null)
+  const [replaceValue, setReplaceValue] = useState(replaceQuery)
+  const [preserveCaseReplace, setPreserveCaseReplace] = useState<boolean>(
+    searchOptions.preserveCaseReplace
+  )
+
+  const textAreaRows = useCallback(() => {
+    const replaceNumLines = replaceValue ? replaceValue.split('\n').length : 0
+    return replaceNumLines == 0 || replaceNumLines == 1
+      ? 1
+      : replaceNumLines + 1
+  }, [replaceValue])
+
+  const updateReplaceValue: ChangeEventHandler<HTMLTextAreaElement> = useCallback(
+    (event) => {
+      setReplaceValue(event.target.value)
+      if (onReplaceQueryChange) {
+        onReplaceQueryChange(event.target.value)
+      }
+    },
+    [onReplaceQueryChange]
+  )
+
+  const handleOnReplaceClose = useCallback(() => {
+    if (onReplaceClose != null) {
+      onReplaceClose()
+    }
+  }, [onReplaceClose])
+
+  const addNewlineToReplaceValue = useCallback(() => {
+    if (replaceTextAreaRef.current === null) {
+      return
+    }
+    const startPosition = replaceTextAreaRef.current.selectionStart
+    const endPosition = replaceTextAreaRef.current.selectionEnd
+    if (startPosition == endPosition) {
+      // no selection -> add newline at this location
+      replaceTextAreaRef.current.setRangeText(
+        '\n',
+        startPosition,
+        endPosition,
+        'end'
+      )
+    } else {
+      // selection -> replace it with newline
+      replaceTextAreaRef.current.setRangeText(
+        '\n',
+        startPosition,
+        endPosition,
+        'end'
+      )
+    }
+    setReplaceValue(replaceTextAreaRef.current.value)
+    if (onReplaceQueryChange) {
+      onReplaceQueryChange(replaceValue)
+    }
+  }, [onReplaceQueryChange, replaceValue])
+
+  const focusReplaceTextAreaInput = useCallback(
+    (replaceValueLength?: number) => {
+      if (replaceTextAreaRef.current == null) {
+        return
+      }
+      replaceTextAreaRef.current.focus()
+      if (replaceValueLength && replaceValueLength > 0) {
+        replaceTextAreaRef.current.selectionEnd += replaceValueLength
+        replaceTextAreaRef.current.selectionStart = replaceValueLength
+      }
+    },
+    []
+  )
+
+  const replaceSingleMarker = useCallback(
+    (marker: TextMarker) => {
+      const markerPos: MarkerRange = marker.find() as MarkerRange
+      if (!markerPos) {
+        return false
+      }
+      const foundMarkText = codeMirror.getRange(markerPos.from, markerPos.to)
+      if (foundMarkText.length >= 1 && preserveCaseReplace) {
+        // if first character was upper case - set the replacement to upper case too
+        codeMirror.replaceRange(
+          foundMarkText.charAt(0) + replaceQuery.substr(1),
+          markerPos.from,
+          markerPos.to,
+          '@ignore'
```

Comment:
It's a hack, using [origin](url) property of the call to send ignore the information - this event emit. 
It is used to ignore updating search results in event handler bound by local search component. 
Since we don't want to do this for each replaceRange call, only at the end of replacing (could be `replaceAll` and a lot of items could be replaced, so it does not make sense to do local search update on each replace)

Reply:
Do you think we can do something like `replaceAll` method in this script? https://codemirror.net/addon/search/search.js
I'd like to avoid using this hack since it doesn't feel like intended behavior.

Reply:
Um, I'll see if it can be done tomorrow, but if `cursor.replace` also triggers change event it should not be possible to avoid the hack.

Reply:
I investigated further:

1. Removed `@ignore` 
2. Tried using `cm.startOperation`, `cm.endOperation` for all operations under `replaceAll`
3. Tried using `cm.operation`
4. Tried using their `replaceAll` function which uses `cm.operation`

All of this still triggers at some point all events (unfortunately, but documentation says it should buffer all events and trigger only 1 update - with `changes.next` which points to next change but this does not happen.

It does thou work better with `cm.operation` in a sense that `replaceAll` functionality works without `@ignore`, but it is not optimized since change event is still fired N times (N - replacements).

As for the hack that works perfectly, I think only problem would be if someone declares change event on codemirror and do not ignore the `origin = @ignore` ones.

And also since documentation says this about the parameter: 
`
/**  String representing the origin of the change event and wether it can be merged with history */
        origin?: string;
`

I think this can be used as the origin of the change event, in this case on replaceAll changes, the origin is "replaceFeature" and it should be ignored since the feature itself will refresh the UI when needed.

Please see if you want me to use a non-optimized version, and operation approach, or if you can investigate why are all events still fired separately.

Reply:
Oh, my mistake, there are two events:

`change` and `changes`, the latter one actually waits and receives operations, all combined from `operation` method.
Change is the one which receives all single operations nevertheless combining.

Went digging into codemirror code and found about all this, including how operation works, read some comments etc.

I'll fix this hack and try to push the new version soon, @Rokt33r

Reply:
The new version pushed, seems to work really nice now.